source/opt/scalar_replacement_pass.cpp

```diff
@@ -172,10 +172,15 @@ bool ScalarReplacementPass::ReplaceWholeDebugDeclare(
   // Add DebugValue instruction with Indexes operand and Deref operation.
   int32_t idx = 0;
   for (const auto* var : replacements) {
+    // We use the nullptr instruction for the scope and line info instruction
+    // of the new DebugValue because it will be used as DebugDeclare, not
+    // DebugValue. Setting the scope and line info must be done when we
+    // distribute DebugValue instructions for a DebugDeclare.
```

Comment:
I do not understand this what you mean by "Setting the scope and line info must be done when we distribute DebugValue instructions for a DebugDeclare".  

This part of the code is replace one instruction with a series of other instructions.  Why wouldn't the new instruction get the scope and line number of the instruction being removed?  Also, we cannot assume that any other passes are being run after this.  Users can pick and choose the optimizations that they want to run.

Reply:
> Why wouldn't the new instruction get the scope and line number of the instruction being removed?

The removing instruction in this case must be `DebugDeclare`. Since `DebugDeclare` must be position-independant under its scope, I thought it should not have `OpLine`. `DebugDeclare` has `DebugLocalVariable` that has the information of the line and the scope.

> we cannot assume that any other passes are being run after this.

I did not mean that we want to reply on other passes for it. I wanted to explain `DebugValue` replaced with the `DebugDeclare` can have the line information.

To make it more clear, I added a function `AddDebugDeclareForScalar()` that adds `DebugValue` with `Deref` operation for the `DebugDeclare` of each component of an aggregate.
It uses the line and the scope of the aggregate's `DebugDeclare` for the debug declares (`DebugValue` with `Deref`) of its components.

Reply:
> Since DebugDeclare must be position-independant under its scope, I thought it should not have OpLine

What do you mean by "you think"?  What does the spec say?  Does the spec allows the DebugDeclare to have and OpLine and DebugScope that apply it? If so, then the new DebugValue should have the same OpLine and DebugScope apply to it.

Reply:
> To make it more clear, I added a function AddDebugDeclareForScalar() that adds DebugValue with Deref operation for the DebugDeclare of each component of an aggregate.

The problem is not that the code is not clear enough.  No extra function is needed.  I am questioning why there the debug info for the instruction is not being copied over.  Saying that you are adding the debug declare for the scalar does not help explain that.

Reply:
Adding `OpLine` for `DebugDeclare` can make a debugger step back and forth if the debugger steps based on `OpLine`s.
For example,
```
void main() {
...
foo(x);  // <-- second stepping goes to here
...
}

void foo(int y) {  // <-- first stepping goes to here
...   // <-- third stepping goes to here
}
```

In particular when `DebugDeclare`s for scalar variables generated by the scalar replacement are shuffled with other instructions, the stepping will become more complicatedly to and fro.

Reply:
> In particular when DebugDeclares for scalar variables generated by the scalar replacement are shuffled with other instructions, the stepping will become more complicatedly to and fro.

Yes that happens when you optimize code, and do any type of code motion.  It does make things more difficult to keep track of what is going on, but it is not something the compiler does wrong.  The compiler preserves the correct information, and then the consumers of the information do what they think is best for their purpose.

Reply:
done